### PR TITLE
tcpdump project

### PR DESCRIPTION
**Pull request by headshog**

### CODE REVIEW

File: projects/tcpdump/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+
+FROM sydr/ubuntu20.04-sydr-fuzz
+
+RUN apt-get update && apt-get install -y make cmake flex bison
```
Comment (kobrineli) [inline projects/tcpdump/Dockerfile:19]: make and cmake are already installed

File: projects/tcpdump/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+RUN git clone https://github.com/the-tcpdump-group/libpcap.git libpcap
+RUN cd libpcap && git checkout e0be22909d285fa721379542a754df696c314da3
+
+RUN git clone https://github.com/the-tcpdump-group/tcpdump.git tcpdump
+RUN cd tcpdump && git checkout 1e63a2f6903de9303931f3135db88932d4e5785a
+
+RUN mkdir corpus && mv tcpdump/tests/*.pcap corpus/
+
+COPY build.sh load_sydr_packet.c /
```
Comment (kobrineli) [inline projects/tcpdump/Dockerfile:29]: Rename target (originally it was `fuzz_pretty_print_packet`)

File: projects/tcpdump/build.sh [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+
+#Build targets for libfuzzer
```
Comment (kobrineli) [inline projects/tcpdump/build.sh:19]: Remove comment

File: projects/tcpdump/load_sydr_packet.c [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+#include <string.h>
+
+static uint64_t packets_captured = 0;
+
+void print_packet(unsigned char *user, const struct pcap_pkthdr *h, const unsigned char *sp) {
+  ++packets_captured;
+  pretty_print_packet((netdissect_options *)user, h, sp, packets_captured);
+}
+
+FILE *bufferToFile(const uint8_t *Data, size_t Size) {
```
Comment (SweetVishnya) [inline projects/tcpdump/load_sydr_packet.c:19]: We should directly read from file in sydr target

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: projects/tcpdump/Dockerfile (hunk 2 above)
Comment (kobrineli) [reply projects/tcpdump/Dockerfile:29]: You can name source `pretty_print_packet`

And targets for different tools you can name like `pretty_..._sydr`, `pretty_..._fuzz`, etc.

File: projects/tcpdump/build.sh [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    export CC="clang"
+    export CFLAGS="-g -fsanitize=fuzzer-no-link,address,bounds,integer,undefined,null,float-divide-by-zero"
+    export TARGET_CFLAGS="-g -fsanitize=fuzzer,address,bounds,integer,undefined,null,float-divide-by-zero"
+    export LINK_FLAGS="$TARGET_CFLAGS -I/libpcap -I/tcpdump ../load_sydr_packet.c -o ../fuzzer_packet\
+                       /libpcap/libpcap.a /tcpdump/libnetdissect.a"
+fi
+
+if [[ $CONFIG = "afl" ]]
+then
+    cd /libpcap && make clean
```
Comment (kobrineli) [inline projects/tcpdump/build.sh:31]: Move it to the common part before `make -j$(nproc)`

File: projects/tcpdump/build.sh [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    cd /tcpdump && make clean
```
Comment (kobrineli) [inline projects/tcpdump/build.sh:32]: Same here and later

File: projects/tcpdump/build.sh [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    export CC="afl-clang-fast"
+    export CFLAGS="-g -fsanitize=address,bounds,integer,undefined,null,float-divide-by-zero"
+    export TARGET_CFLAGS="-g -fsanitize=fuzzer,address,bounds,integer,undefined,null,float-divide-by-zero"
```
Comment (kobrineli) [inline projects/tcpdump/build.sh:35]: Do we need `TARGET_CFLAGS` be separate from `LINK_FLAGS`?

File: projects/tcpdump/load_sydr_packet.c [hunk 8]
```diff
@@ -0,0 +1,91 @@
+#include <pcap/pcap.h>
```
Comment (kobrineli) [inline projects/tcpdump/load_sydr_packet.c:1]: Add license

**Review by kobrineli — CHANGES_REQUESTED** [review]

**Review by kobrineli — CHANGES_REQUESTED** [review]
Add README.md

File: projects/tcpdump/load_sydr_packet.c (hunk 4 above)
Comment (headshog) [reply projects/tcpdump/load_sydr_packet.c]: done

File: projects/tcpdump/Dockerfile (hunk 1 above)
Comment (headshog) [reply projects/tcpdump/Dockerfile]: done

File: projects/tcpdump/Dockerfile (hunk 2 above)
Comment (headshog) [reply projects/tcpdump/Dockerfile]: done

File: projects/tcpdump/build.sh (hunk 5 above)
Comment (headshog) [reply projects/tcpdump/build.sh]: done

File: projects/tcpdump/build.sh (hunk 6 above)
Comment (headshog) [reply projects/tcpdump/build.sh]: done

File: projects/tcpdump/build.sh (hunk 7 above)
Comment (headshog) [reply projects/tcpdump/build.sh]: removed

File: projects/tcpdump/README.md [hunk 9]
```diff
@@ -0,0 +1,42 @@
+# tcpdump
+
+tcpdump is a data-network packet analyzer computer program that runs under a command line interface.
```
Comment (kobrineli) [inline projects/tcpdump/README.md:3]: runs -> is run

File: projects/tcpdump/load_sydr_packet.c (hunk 8 above)
Comment (headshog) [reply projects/tcpdump/load_sydr_packet.c]: done

**Comment by headshog** [comment]
> Add README.md

added

File: projects/tcpdump/README.md [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+## Build Docker
+
+    $ sudo docker build -t oss-sydr-fuzz-tcpdump .
+
+## Run Hybrid Fuzzing
+
+Unzip Sydr (`sydr.zip`) in `projects/libpcap` directory:
```
Comment (kobrineli) [inline projects/tcpdump/README.md:11]: `libpcap` -> `tcpdump`

File: projects/tcpdump/build.sh (hunk 3 above)
Comment (headshog) [reply projects/tcpdump/build.sh]: done

**Review by kobrineli — CHANGES_REQUESTED** [review]

File: projects/tcpdump/pretty_print_packet.c [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+  ndo->ndo_tflag++;
+  ndo->ndo_uflag++;
+  ndo->ndo_vflag++;
+  ndo->ndo_xflag++;
+  ndo->ndo_Xflag++;
+  ndo->ndo_Aflag++;
+  ndo->ndo_Hflag++;
+
+#ifdef SYDR
+  fd = fopen(Data, "r");
```
Comment (SweetVishnya) [inline projects/tcpdump/pretty_print_packet.c:87]: something wrong with indentation

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: projects/tcpdump/pretty_print_packet.c (hunk 11 above)
Comment (headshog) [reply projects/tcpdump/pretty_print_packet.c]: changed

File: projects/tcpdump/pretty_print_packet.c [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+	ndo->ndo_Aflag++;
+	ndo->ndo_Hflag++;
+
+#ifdef SYDR
+	fd = fopen(Data, "r");
+#else
+	fd = bufferToFile(Data, Size);
+#endif
+ 	if (fd == NULL)
+				return 0;
```
Comment (SweetVishnya) [inline projects/tcpdump/pretty_print_packet.c:92]: bad indentation

File: projects/tcpdump/pretty_print_packet.c [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+static uint64_t packets_captured = 0;
+
+void print_packet(unsigned char *user, const struct pcap_pkthdr *h, const unsigned char *sp) {
+	++packets_captured;
+	pretty_print_packet((netdissect_options *)user, h, sp, packets_captured);
+}
+
+FILE *bufferToFile(const uint8_t *Data, size_t Size) {
+	FILE *fd;
+	fd = tmpfile();
```
Comment (SweetVishnya) [inline projects/tcpdump/pretty_print_packet.c:37]: shouldn't we remove temporary file later?

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: projects/tcpdump/pretty_print_packet.c (hunk 13 above)
Comment (headshog) [reply projects/tcpdump/pretty_print_packet.c:37]: I found this in `man tmpfile`: The tmpfile() function opens a unique temporary file in binary read/write (w+b) mode.  The file will be automatically deleted when it is closed or the program terminates.

File: projects/tcpdump/pretty_print_packet.c (hunk 12 above)
Comment (headshog) [reply projects/tcpdump/pretty_print_packet.c]: fixed

File: projects/tcpdump/pretty_print_packet.c [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+#else
+	fd = bufferToFile(Data, Size);
+#endif
+ 	if (fd == NULL)
+		return 0;
+
+	pkts = pcap_fopen_offline(fd, eBuf);
+	if (pkts == NULL) {
+			fclose(fd);
+			return 0;
```
Comment (SweetVishnya) [inline projects/tcpdump/pretty_print_packet.c:97]: bad indentation: check github diff

Comment (headshog) [reply projects/tcpdump/pretty_print_packet.c]: fixed

**Review by SweetVishnya — APPROVED** [review]

**Pull request merged**